From: Don Howard <dhoward@redhat.com>
Date: Tue, 3 May 2011 15:36:56 -0400
Subject: [net] bridge/netfilter: fix ebtables information leak
Message-id: <1304437016-8902-1-git-send-email-dhoward@redhat.com>
Patchwork-id: 35744
O-Subject: [RHEL5.7 PATCH] bridge: netfilter: fix information leak
Bugzilla: 681326
CVE: CVE-2011-1080
RH-Acked-by: Don Zickus <dzickus@redhat.com>
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>

Backport from upstream.
Fixes bz 681326, CVE-2011-1080
Brew build: https://brewweb.devel.redhat.com/taskinfo?taskID=3289598
No regressions seen in rhts testing.

commit d846f71195d57b0bbb143382647c2c6638b04c5a
Author: Vasiliy Kulikov <segoon@openwall.com>
Date:   Mon Feb 14 16:49:23 2011 +0100

    bridge: netfilter: fix information leak

    Struct tmp is copied from userspace.  It is not checked whether
    the "name" field is NULL terminated.  This may lead to buffer
    overflow and passing contents of kernel stack as a module name to
    try_then_request_module() and, consequently, to modprobe
    commandline.  It would be seen by all userspace processes.

    Signed-off-by: Vasiliy Kulikov <segoon@openwall.com>
    Signed-off-by: Patrick McHardy <kaber@trash.net>

Signed-off-by: Don Howard <dhoward@redhat.com>

Index: kernel/net/bridge/netfilter/ebtables.c
===================================================================
--- kernel.orig/net/bridge/netfilter/ebtables.c
+++ kernel/net/bridge/netfilter/ebtables.c
@@ -942,6 +942,8 @@ static int do_replace(void __user *user,
 	if (tmp.num_counters >= INT_MAX / sizeof(struct ebt_counter))
 		return -ENOMEM;
 
+	tmp.name[sizeof(tmp.name) - 1] = 0;
+
 	countersize = COUNTER_OFFSET(tmp.nentries) * 
 					(highest_possible_processor_id()+1);
 	newinfo = vmalloc(sizeof(*newinfo) + countersize);
